LaTeX/TeX.sublime-syntax

```diff
@@ -111,6 +118,54 @@ contexts:
       captures:
         1: punctuation.definition.backslash.tex
 
+  registers:
+    - match: (\\)({{registers}}){{endcs}}
+      captures:
+        1: punctuation.definition.backslash.tex
+        2: storage.type.tex
+      push: register-id
+
+  register-id:
+    - meta_scope: meta.register.tex
+    - match: \d+
+      scope: meta.number.integer.decimal constant.numeric.value.tex
+      pop: 1
+    - match: (\\){{cmdname}}
+      captures:
+        1: punctuation.definition.backslash.tex
+      pop: 1
+    - include: paragraph-pop
+    - include: else-pop
+
+  register-allocations:
+    - match: (\\)new{{registers}}{{endcs}}
+      scope: keyword.declaration.register.tex storage.modifier.register.tex
+      captures:
+        1: punctuation.definition.backslash.tex
+      push: def-registername
+
+  def-registername:
+    - match: (\\){{cmdname}}
+      scope: entity.name.constant.tex
+      captures:
+        1: punctuation.definition.backslash.tex
+      pop: 1
+
+    - match: \{
+      scope: punctuation.definition.group.brace.begin.tex
+      push:
+        - meta_scope: meta.group.brace.latex
+        - match: \}
+          scope: punctuation.definition.group.brace.end.tex
+          pop: 2
+        - match: (\\){{cmdname}}
+          scope: entity.name.constant.tex
+          captures:
+            1: punctuation.definition.backslash.tex
+
+    - include: paragraph-pop
+    - include: else-pop
+
```

Comment:
1. I'd suggest to try to avoid anonymous contexts in new/rewritten code in order to improve support for inheritance. We should therefore always try to directly push a named context (see: `inside-register-name-group`). Syntaxes use different kinds of naming schemes one of them is to name contexts `inside-` if they describe a part inside of brackets. Another scheme is to just append `-body` (e.g.: `register-name-group-body`). We should decide for one and go with it.

2. Maybe a stupid habbit of no practical use, but maybe we should try to organice bracket contexts (and pushing ones) first, if possible. Hence I'd suggest to move `- match: \{` to the beginning. Just to have some kind of highlevel code structure strategy.

```suggestion
    - match: \{
      scope: punctuation.definition.group.brace.begin.tex
      push: inside-register-name-group
    - match: (\\){{cmdname}}
      scope: entity.name.constant.tex
      captures:
        1: punctuation.definition.backslash.tex
      pop: 1
    - include: paragraph-pop
    - include: else-pop

  inside-register-name-group:
    - meta_scope: meta.group.brace.latex
    - match: \}
      scope: punctuation.definition.group.brace.end.tex
      pop: 2
    - match: (\\){{cmdname}}
      scope: entity.name.constant.tex
      captures:
        1: punctuation.definition.backslash.tex

```

Reply:
I'm not sure if the inheritance argument makes much sense here, in that it seems very unlikely that there'd be a situation in which you'd change `inside-register-name-group` but not `def-registername` itself, but its probably best to be consistent with these things.

Reply:
All syntaxes move towards the the goal to only use named contexts. Transition should be made as we move forwared with other things as well. Maybe create dedicated PRs for such restructurings as well, such as #3532.

We should not make any assumptions about possibilities or likelyness of contexts being inherit or not as a base syntax can't know that for sure.

For consistency reasons all contexts should be named at some point so all are available for inheritance if required.

Reply:
Just as an example: Assume LaTeX would allow more complex expressions in braced register names.

It would need to rewrite the whole `def-registername` context to do with your code. Using named contexts for `inside-register-name-group` enables it to inject such patterns into braces without touching more than really needed.

Reply:
So would you also argue that there should not be more than one `match` pattern (except maybe else-pop) per context? Because one could have (almost equivalently) defined this context also as matching either
`{{cmdname}}` or `\{\s*{{cmdname}\s*\}`, in which case the discussion about anonymous contexts would not have come up, but the situation (a derived syntax might want to change the stuff in the braces) would be the same.

Reply:
The smaller contexts are the more fine gained an inheriting syntax can override single aspects. I wouldn't however argue to reduce all contexts to a single pattern. We could do that, but at some point we run out of names 😆 - or they would get insane long, maybe.

Patterns like `\{\s*{{cmdname}\s*\}` are often a relic of old TextMate syntaxes. In case such expressions are well known to be used in single line statements only, we could probably keep them as is, but there are several compelling reasons not to do so:

1. if a syntax supports them to span multiple lines, I'd always push dedicated contexts. It would help not breaking highlighting in such scenarios - even if they appear temporarily while editing content.

2. pushing/popping contexts is cheaper than huge patterns with several capture groups, which increases overall performance of a syntax.

3. UX with regards to incomplete code is often better. Just imagine `{}` being typed without the context name. Braces wouldn't be highlighted with capture groups, but they are with pushing contexts.  
   Basically I'd compare braces to quotation marks in strings here. All strings push a context onto stack and so all those braces should do as well, IMHO.

4. readability, debugging may become easiear as named contexts are listed in the "show scope" popup. _(targets named contexts)_

5. supporting inheritance (or providing more flexibility to support it) is the most recent argument added to that list. _(targets named contexts)_

Reply:
Totally forgot about variable interpolation. If a syntax wants to use TeX/LaTeX as template, it might be useful to be able to inject variable interpolations.